Feat: Add N-D parallelism docs (#2989)
OpenAccess-AI-Collective/axolotl: docs/nd_parallelism.qmd

# N-D Parallelism

> Axolotl enables training large models via composable parallelism methods, primarily leveraging Pytorch's `FSDP` and `DeviceMesh`. This is essential when:

Reviewer: ```suggestion
Axolotl enables training models at scale by composing different parallelism techniques. This is essential when:
```

- A model's weights are too large to fit on a single GPU's memory.
- A model's activations, especially with very long contexts, are too large for a single GPU.
- You want to accelerate training by using multiple GPUs or nodes.

or combinations of the above!

## Core Concepts

Parallelism strategies can be combined. The key is understanding how each one divides the workload. PyTorch's `DeviceMesh` is the modern way to manage these combinations, creating a logical grid of your GPUs and assigning different parallel strategies to different dimensions of the grid.

> ### Data Parallelism (DP & FSDP)

Reviewer: ```suggestion
### Data Parallelism and Fully Sharded Data Parallelism (DP & FSDP)
```

Data Parallelism focuses on splitting the global data batch across GPUs.

> - Distributed Data Parallel (DDP): The classic approach. The full model is replicated on every GPU. Each GPU processes a different slice of the data batch. Gradients are averaged across all GPUs (`all_reduce`) after the backward pass to keep the models synchronized. This is simple but memory-inefficient, as each GPU must hold the entire model, its gradients, and optimizer states.

Reviewer: ```suggestion
- Distributed Data Parallel (DDP): The classic approach. The full model is replicated on every GPU. Each GPU processes a different slice of the data batch. Gradients are then averaged across all GPUs after the backward pass to keep the models synchronized. This can substantially improve data throughput compared to single-device training, but requires that each GPU is able to hold the entire model, its gradients, and optimizer states.
```

- [Fully Sharded Data Parallel (FSDP)](multi-gpu.qmd#fully-sharded-data-parallel-(fsdp)): A highly memory-efficient form of data parallelism (inspired by DeepSpeed's ZeRO). Instead of replicating the model, FSDP shards the model's *parameters, gradients, and optimizer states* across the GPUs in the data-parallel group. During computation, each GPU receives the specific parameters it needs via an `all_gather` operation just before they are used, and they can be discarded immediately after (`reshard-after-forward`).
    - FSDP maps to ZeRO stages:
        - ZeRO-2 (`reshard_after_forward=False`): Shards gradients and optimizer states. Model weights are replicated on each GPU.
        - ZeRO-3 (`reshard_after_forward=True`): Shards gradients, optimizer states, AND model parameters. This provides the most memory savings at the cost of more communication (re-gathering parameters for both forward and backward passes).

> ### Tensor Parallelism (TP)

Reviewer: Let's add a big note saying that TP is an experimental feature

| --- | :---: | :---: |:---:|:---:|---|
| **FSDP** (ZeRO-3) | 1 | >1 | 1 | 1 | ✅ Fully supported. Shards model across all GPUs. |
| **HSDP** | >1 | >1 | 1 | 1 | ✅ Fully supported. FSDP intra-node, DDP inter-node. |
| **FSDP + TP** | 1 | >1 | >1 | 1 | ✅ **2D Parallelism**. Shards the model across a `dp_shard` group, and TP-splits layers within the `tp` group. |
| **HSDP + TP** | >1 | >1 | >1 | 1 | ✅ **3D Parallelism**. A powerful but complex combination. |
| **FSDP + CP** | 1 | >1 | 1 | >1 | ✅ **2D Parallelism**. Combines FSDP with context parallelism. |
| **FSDP + TP + CP**| 1 | >1 | >1| >1| ✅ **3D Parallelism**. Another advanced combination. |
| DDP + TP/CP | >1 | 1 | >1 | >1 | ❌ **Not Supported**. The `ParallelismConfig` explicitly prevents this, as composing pure DDP with TP/CP without FSDP is inefficient and complex. You should use FSDP instead (`dp_shard_size > 1`). |
| Just TP / CP | 1 | 1 | >1 | >1 | ✅ Supported. Useful for inference or when the model fits on one GPU but context is too long. |

- `tp_size` refers to `tensor_parallel_size`
- `cp_size` refers to `context_parallel_size`

> 

Reviewer: Let's also indicate what dp_replicate and dp_shard mean

Author: I moved Usage section up which has explanations for dp_replicate_size and dp_shard_size. The reason I mentioned those two here was because they were abbreviated in the table header.

## Usage

> Axolotl will construct the appropriate `DeviceMesh` for PyTorch depending on the following configs.

Reviewer: I think we can omit this

```yaml
# The number of GPUs to shard the model parameters across (FSDP dimension).

> dp_shard_size: 4

Reviewer: fsdp_config must be specified here

# The number of times to replicate the sharded model (DDP dimension).
dp_replicate_size: 2

# Number of GPUs for Tensor Parallelism.
tensor_parallel_size: 1  # (default is 1, no TP)

# Number of GPUs for Context/Sequence Parallelism.
context_parallel_size: 1 # (default is 1, no CP)
```

Note: We recommend FSDP. DeepSpeed is only compatible with `tensor_parallel_size`.

## Examples:

1.  HSDP on 2 nodes with 4 GPUs each (8 GPUs total):
    - You want FSDP within each node and DDP across nodes.
    - Set `dp_shard_size: 4` and `dp_replicate_size: 2`.

2.  FSDP + TP on a single 8-GPU node:
    - You want to split the model across 4 GPUs using FSDP, and further split each layer across 2 GPUs with TP.

>     - Set `dp_shard_size: 4`, `tensor_parallel_size: 2`, `dp_replicate_size: 1`.

Reviewer: ```suggestion
    - Set `dp_shard_size: 4` and `tensor_parallel_size: 2`.
```
